src/Authoring/WinRT.SourceGenerator/Generator.cs

```diff
@@ -248,9 +196,54 @@ public void Execute(GeneratorExecutionContext context)
             Logger.Log("Done");
             Logger.Close();
         }
+    }
 
-        public void Initialize(GeneratorInitializationContext context)
-        {
-        }
+    [Generator]
+    public class SourceGenerator : ISourceGenerator
+    {
+        public void Execute(GeneratorExecutionContext context)
+        {
+            if (!context.IsCsWinRTComponent())
+            {
+                return;
+            }
+
+            ComponentGenerator generator = new ComponentGenerator(context);
+            generator.Generate();
+        }
+
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new WinRTSyntaxReciever());
+        }
+    }
+
+    class WinRTSyntaxReciever : ISyntaxReceiver
+    {
+        public List<MemberDeclarationSyntax> Declarations = new List<MemberDeclarationSyntax>();
+
+        public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
+        {
+            if (syntaxNode is not MemberDeclarationSyntax decaralation || !IsPublic(decaralation))
+            {
+                return;
+            }
+
+            if (syntaxNode is ClassDeclarationSyntax ||
+                syntaxNode is InterfaceDeclarationSyntax ||
+                syntaxNode is EnumDeclarationSyntax ||
+                syntaxNode is DelegateDeclarationSyntax ||
+                syntaxNode is StructDeclarationSyntax ||
+                syntaxNode is NamespaceDeclarationSyntax)
+            {
+                Declarations.Add(decaralation);
+            }
+        }
+
+        private bool IsPublic(MemberDeclarationSyntax member)
+        {
+            // We detect whether partial types are public using symbol information later.
+            return member.Modifiers.Any(m => m.IsKind(SyntaxKind.PublicKeyword) || m.IsKind(SyntaxKind.PartialKeyword));

```

Comment:
can you have a `internal partial class MyClass` ?

Reply:
Yes you can which will be caught by the DeclaredAccessibility check done later on using the symbol information.